File: content/snippets/cpp/interview_lcci_04_02_minimum_height_tree.cpp
#include <iostream>
#include <vector>

struct Node { int v; Node* l; Node* r; Node(int x): v(x), l(nullptr), r(nullptr) {} };

Node* build(const std::vector<int>& a, int lo, int hi) {
    if (lo > hi) return nullptr;
    int mid = (lo + hi) / 2;
    Node* root = new Node(a[mid]);
    root->l = build(a, lo, mid - 1);
    root->r = build(a, mid + 1, hi);
    return root;
}

void inorder(Node* n) { if (!n) return; inorder(n->l); std::cout << n->v << ' '; inorder(n->r); }

int main() {
    std::vector<int> nums{1,2,3,4,5,6,7};
    Node* root = build(nums, 0, static_cast<int>(nums.size()) - 1);
    inorder(root);
    std::cout << '\n';
}

File: content/snippets/go/interview_lcci_04_02_minimum_height_tree.go
package main

import "fmt"

type Node struct {
	Val         int
	Left, Right *Node
}

func build(a []int, lo int, hi int) *Node {
	if lo > hi {
		return nil
	}
	mid := (lo + hi) / 2
	root := &Node{Val: a[mid]}
	root.Left = build(a, lo, mid-1)
	root.Right = build(a, mid+1, hi)
	return root
}

func inorder(n *Node, out *[]int) {
	if n == nil {
		return
	}
	inorder(n.Left, out)
	*out = append(*out, n.Val)
	inorder(n.Right, out)
}

func main() {
	nums := []int{1, 2, 3, 4, 5, 6, 7}
	root := build(nums, 0, len(nums)-1)
	out := []int{}
	inorder(root, &out)
	for i, v := range out {
		if i > 0 {
			fmt.Print(" ")
		}
		fmt.Print(v)
	}
	fmt.Println()
}
